File: Problems/Algorithms/2. Add Two Numbers/AddTwoNumbers.java
/**
 * Definition for singly-linked list.
 * public class ListNode {
 *     int val;
 *     ListNode next;
 *     ListNode() {}
 *     ListNode(int val) { this.val = val; }
 *     ListNode(int val, ListNode next) { this.val = val; this.next = next; }
 * }
 */
class Solution {
    public ListNode addTwoNumbers(ListNode l1, ListNode l2) {
        ListNode ans = new ListNode(0);
        ListNode node1 = l1;
        ListNode node2 = l2;
        ListNode node3 = ans;
        int carry = 0;
        
        while (node1 != null && node2 != null) {
            int sum = node1.val + node2.val + carry;
            node3.val = sum % 10;
            carry = sum / 10;
            
            if (node1.next == null && node2.next == null) {
                if (carry == 1) {
                    node3.next = new ListNode(1);
                }
                return ans;
            }
            
            node3.next = new ListNode(0);
            node3 = node3.next;
            node1 = node1.next;
            node2 = node2.next;
        }
        
        while (node1 != null) {
            int sum = node1.val + carry;
            node3.val = sum % 10;
            carry = sum / 10;
            
            if (node1.next == null) {
                if (carry == 1) {
                    node3.next = new ListNode(1);
                }
                return ans;
            }
            
            node3.next = new ListNode(0);
            node3 = node3.next;
            node1 = node1.next;
        }
        
        while (node2 != null) {
            int sum = node2.val + carry;
            node3.val = sum % 10;
            carry = sum / 10;
            
            if (node2.next == null) {
                if (carry == 1) {
                    node3.next = new ListNode(1);
                }
                return ans;
            }
            
            node3.next = new ListNode(0);
            node3 = node3.next;
            node2 = node2.next;
        }
        
        return ans;
    }
}

File: Explore/Learn/Introduction to Data Structure/Linked List/Add Two Numbers/AddTwoNumbers.kt
/**
 * Example:
 * var li = ListNode(5)
 * var v = li.`val`
 * Definition for singly-linked list.
 * class ListNode(var `val`: Int) {
 *     var next: ListNode? = null
 * }
 */
class Solution {
    fun addTwoNumbers(l1: ListNode?, l2: ListNode?): ListNode? {
        var ans: ListNode? = ListNode(0)
        var node1: ListNode? = l1
        var node2: ListNode? = l2
        var node3: ListNode? = ans
        var carry: Int = 0
        
        while (node1 != null && node2 != null) {
            var sum: Int = node1!!.`val` + node2!!.`val` + carry
            node3?.`val` = sum % 10
            carry = sum / 10
            
            if (node1?.next == null && node2?.next == null) {
                if (carry == 1) {
                    node3?.next = ListNode(1)
                }
                return ans
            }
            
            node3?.next = ListNode(0)
            node3 = node3?.next
            node1 = node1?.next
            node2 = node2?.next
        }
        
        while (node1 != null) {
            var sum: Int = node1!!.`val` + carry
            node3?.`val` = sum % 10
            carry = sum / 10
            
            if (node1?.next == null) {
                if (carry == 1) {
                    node3?.next = ListNode(1)
                }
                return ans
            }
            
            node3?.next = ListNode(0)
            node3 = node3?.next
            node1 = node1?.next
        }
        
        while (node2 != null) {
            var sum: Int = node2!!.`val` + carry
            node3?.`val` = sum % 10
            carry = sum / 10
            
            if (node2?.next == null) {
                if (carry == 1) {
                    node3?.next = ListNode(1)
                }
                return ans
            }
            
            node3?.next = ListNode(0)
            node3 = node3?.next
            node2 = node2?.next
        }
        
        return ans
    }
}

